doc: configuration and folder cleanup

gmarull opened the pull request:
Prefix Sphinx related folders (static, scripts, extensions...) with an underscore, same as it was done for `_templates`. This emphasizes that these folders do not contain actual documentation. `_static` content is also sorted into categories (js, css...).

I have also simplified/cleaned the `conf.py` file.

gmarull committed:
doc: move static folder to _static

Move static content to _static. It is better to prefix all Sphinx
related folders with an underscore to emphasize that the folder does not
contain actual documentation. Static content gets copied into _static
folder when documentation is built, so it also makes things more
consistent.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move scripts to _scripts

Move Sphinx related content to folders with underscore prefix.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move extensions to _extensions

Move all Sphinx related content to folders prefixed with an underscore.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move all js files to _static/js

Move all Javascript files to the _static/js subfolder.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move all css files to _static/css

Move all CSS files to the _static/css subfolder.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Looks like compliance check doesn't allow empty `__init__.py` files (were added previous to these changes...). This is a valid case and a frequent situation in Python code. cc: @galak @nashif

gmarull committed:
doc: remove commented options from config

Remove options that are commented out from conf.py. It improves
readability of the file.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: remove Sphinx defaults

Remove settings that are Sphinx defaults in order to simplify
configuration.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move redirects to a utility module

Isolate the list of redirects in order to improve conf.py readability.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: remove settings for unused builders

manpage, texinfo and htmlhelp builders are not used, so remove their
settings as they are redundant.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: add configuration reference and remove parameter details

It is easy to have out-of-date comments since Sphinx settings evolve
(defaults, behavior...). A reference to the official documentation is
provided at the top of the file instead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: simplify some html_context logic

Simplify is_release and docs_title logic and move it closer to where it
is used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move imports to the top of the file

Some are not really at the top because of path inserts, but are as up as
possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: organize configuration sections better

Organize configuration sections per group (project, general, HTML,
plugin options...).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: remove unicode prefix from strings

This is not needed on Python 3.x as it's the default.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: use regex to parse Zephyr version

Simplify the logic of version extraction by using regexes.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: use pathlib for paths in configuration file

Switch to pathlib to make configuration more readable. sys.exit has been
replaced with an exception so that in case of failure error is displayed
(error details were hidden).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: format configuration file

Format the configuration file so that it is more consistent (e.g.
uniform usage of single/double quotes).

Some minor cleanup performed too: removed some comments and a default
setting.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlescufi reviewed (commented):
Very nice cleanup. I am just not sure about the underscores before the folders. From what you write in the commit messages, I don't see that they are required at all, and we have no precedent in the tree with leading underscores on folder names. I understand the point of making folders that do not contain actual doc more visible, but could we achieve that instead in some other way?

carlescufi commented:
> Looks like compliance check doesn't allow empty `__init__.py` files (were added previous to these changes...). This is a valid case and a frequent situation in Python code. cc: @galak @nashif

Just add a comment like:
`# File intentionally empty`
inside the file?

gmarull commented:
> Very nice cleanup. I am just not sure about the underscores before the folders. From what you write in the commit messages, I don't see that they are required at all, and we have no precedent in the tree with leading underscores on folder names. I understand the point of making folders that do not contain actual doc more visible, but could we achieve that instead in some other way?

I'd say it is a fairly common practice in Sphinx. We actually do it for templates (we have `_templates` for Sphinx template customizations and `templates` for some documentation content template). This way there is a clear split between what are docs and Sphinx related files.

gmarull committed:
doc: add comment to empty Python files

Explicitely mention that these files are intentionally empty. These
changes should make compliance checks pass.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic commented on doc/_extensions/lexer/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2021 Zephyr Project members and individual contributors
```
Not Nordic?

gmarull replied:
These are just comments to pass compliance checks. As the files were already there I thought it was better to add this copyright message.

nashif replied:
i guess nobody wants to own copyright on an empty file :-D

mbolivar-nordic reviewed (approved):
Great cleanup!

carlescufi commented:
> > Very nice cleanup. I am just not sure about the underscores before the folders. From what you write in the commit messages, I don't see that they are required at all, and we have no precedent in the tree with leading underscores on folder names. I understand the point of making folders that do not contain actual doc more visible, but could we achieve that instead in some other way?
> 
> I'd say it is a fairly common practice in Sphinx. We actually do it for templates (we have `_templates` for Sphinx template customizations and `templates` for some documentation content template). This way there is a clear split between what are docs and Sphinx related files.

Fair enough.
